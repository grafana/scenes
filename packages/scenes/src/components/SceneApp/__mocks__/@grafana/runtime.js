const React = require('react');
const runtime = jest.requireActual('@grafana/runtime');

runtime.PluginPage = jest.fn().mockImplementation((props) => {
  return React.createElement('div', null, props.children);
});

runtime.getPluginImportUtils = () => ({
  getPanelPluginFromCache: jest.fn(() => pluginToLoad),
});

module.exports = runtime;
