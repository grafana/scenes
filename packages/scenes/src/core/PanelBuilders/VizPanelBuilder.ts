import { VizPanel, VizPanelState } from '../../components/VizPanel/VizPanel';
import { DeepPartial } from '../types';
import { FieldConfigBuilder } from './FieldConfigBuilder';
import { FieldConfigOverridesBuilder } from './FieldConfigOverridesBuilder';
import { PanelOptionsBuilder } from './PanelOptionsBuilder';
import { StandardFieldConfig, StandardFieldConfigInterface } from './types';

export class VizPanelBuilder<TOptions extends {}, TFieldConfig extends {}>
  implements StandardFieldConfigInterface<StandardFieldConfig, VizPanelBuilder<TOptions, TFieldConfig>, 'set'>
{
  private _state: VizPanelState<TOptions, TFieldConfig> = {} as VizPanelState<TOptions, TFieldConfig>;
  private _fieldConfigBuilder: FieldConfigBuilder<TFieldConfig>;
  private _panelOptionsBuilder: PanelOptionsBuilder<TOptions>;

  public constructor(
    pluginId: string,
    pluginVersion: string,
    defaultOptions?: () => Partial<TOptions>,
    defaultFieldConfig?: () => TFieldConfig
  ) {
    this._state.title = '';
    this._state.description = '';
    this._state.displayMode = 'default';
    this._state.hoverHeader = false;
    this._state.pluginId = pluginId;
    this._state.pluginVersion = pluginVersion;

    this._fieldConfigBuilder = new FieldConfigBuilder(defaultFieldConfig);
    this._panelOptionsBuilder = new PanelOptionsBuilder(defaultOptions);
  }

  /**
   * Set panel title.
   */
  public setTitle(title: VizPanelState['title']): this {
    this._state.title = title;
    return this;
  }

  /**
   * Set panel description.
   */
  public setDescription(description: VizPanelState['description']): this {
    this._state.description = description;
    return this;
  }

  /**
   * Set panel display mode.
   */
  public setDisplayMode(displayMode: VizPanelState['displayMode']): this {
    this._state.displayMode = displayMode;
    return this;
  }

  /**
   * Set if panel header should be shown on hover.
   */
  public setHoverHeader(hoverHeader: VizPanelState['hoverHeader']): this {
    this._state.hoverHeader = hoverHeader;
    return this;
  }

  /**
   * Set panel menu scene object.
   */
  public setMenu(menu: VizPanelState['menu']): this {
    this._state.menu = menu;
    return this;
  }

  /**
   * Set scene object or react component to use as panel header actions.
   */
  public setHeaderActions(headerActions: VizPanelState['headerActions']): this {
    this._state.headerActions = headerActions;
    return this;
  }

  /**
   * Set color.
   */
  public setColor(color: StandardFieldConfig['color']): this {
    this._fieldConfigBuilder.setColor(color);
    return this;
  }

  /**
   * Set number of decimals to show.
   */
  public setDecimals(decimals: StandardFieldConfig['decimals']): this {
    this._fieldConfigBuilder.setDecimals(decimals);
    return this;
  }

  /**
   * Set field display name.
   */
  public setDisplayName(displayName: StandardFieldConfig['displayName']): this {
    this._fieldConfigBuilder.setDisplayName(displayName);
    return this;
  }

  /**
   * Set the standard field config property filterable.
   */
  public setFilterable(filterable: StandardFieldConfig['filterable']): this {
    this._fieldConfigBuilder.setFilterable(filterable);
    return this;
  }

  /**
   * Set data links.
   */
  public setLinks(links: StandardFieldConfig['links']): this {
    this._fieldConfigBuilder.setLinks(links);
    return this;
  }

  /**
   * Set value mappings.
   */
  public setMappings(mappings: StandardFieldConfig['mappings']): this {
    this._fieldConfigBuilder.setMappings(mappings);
    return this;
  }

  /**
   * Set the standard field config property max.
   */
  public setMax(max: StandardFieldConfig['max']): this {
    this._fieldConfigBuilder.setMax(max);
    return this;
  }

  /**
   * Set the standard field config property min.
   */
  public setMin(min: StandardFieldConfig['min']): this {
    this._fieldConfigBuilder.setMin(min);
    return this;
  }

  /**
   * Set the standard field config property noValue.
   */
  public setNoValue(noValue: StandardFieldConfig['noValue']): this {
    this._fieldConfigBuilder.setNoValue(noValue);
    return this;
  }

  /**
   * Set the standard field config property thresholds.
   */
  public setThresholds(thresholds: StandardFieldConfig['thresholds']): this {
    this._fieldConfigBuilder.setThresholds(thresholds);
    return this;
  }

  /**
   * Set the standard field config property unit.
   */
  public setUnit(unit: StandardFieldConfig['unit']): this {
    this._fieldConfigBuilder.setUnit(unit);
    return this;
  }

  public setCustomFieldConfig<T extends TFieldConfig, K extends keyof T>(id: K, value: DeepPartial<T[K]>): this {
    this._fieldConfigBuilder.setCustomFieldConfig(id, value);
    return this;
  }

  public setOverrides(builder: (b: FieldConfigOverridesBuilder<TFieldConfig>) => void): this {
    this._fieldConfigBuilder.setOverrides(builder);
    return this;
  }

  /**
   * Set an individual panel option. This will merge the value with the existing options.
   */
  public setOption<T extends TOptions, K extends keyof T>(id: K, value: DeepPartial<T[K]>): this {
    this._panelOptionsBuilder.setOption(id, value);
    return this;
  }

  /**
   * Set data provider for the panel.
   */
  public setData(data: VizPanelState['$data']): this {
    this._state.$data = data;
    return this;
  }

  /**
   * Set time range for the panel.
   */
  public setTimeRange(timeRange: VizPanelState['$timeRange']): this {
    this._state.$timeRange = timeRange;
    return this;
  }

  /**
   * Set variables for the panel.
   */
  public setVariables(variables: VizPanelState['$variables']): this {
    this._state.$variables = variables;
    return this;
  }

  /**
   * Set behaviors for the panel.
   */
  public setBehaviors(behaviors: VizPanelState['$behaviors']): this {
    this._state.$behaviors = behaviors;
    return this;
  }

  /**
   * Build the panel.
   */
  public build() {
    const panel = new VizPanel<TOptions, TFieldConfig>({
      ...this._state,
      options: this._panelOptionsBuilder.build(),
      fieldConfig: this._fieldConfigBuilder.build(),
    });

    return panel;
  }
}
