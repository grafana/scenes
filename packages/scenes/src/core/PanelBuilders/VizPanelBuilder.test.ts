import { MappingType, ThresholdsMode } from '@grafana/schema';
import { PanelBuilders } from './index';
import { VizPanelBuilder } from './VizPanelBuilder';

interface OptionsTest {
  numeric: number;
  text: string;
  complex: {
    a: number;
    b: string;
  };
}

interface FieldConfigTest {
  numeric: number;
  text: string;
  complex: {
    a: number;
    b: string;
  };
}

const defaultOption = {
  numeric: 1,
  text: 'text',
  complex: {
    a: 1,
    b: 'text',
  },
};

const createDefaultOptions = (): OptionsTest => defaultOption;

const defaultFieldConfig = {
  numeric: 1,
  text: 'text',
  complex: {
    a: 1,
    b: 'text',
  },
};
const createDefaultFieldConfig = (): FieldConfigTest => defaultFieldConfig;

const getTestBuilder = () =>
  new VizPanelBuilder<OptionsTest, FieldConfigTest>(
    'testplugin',
    '1.0.0',
    createDefaultOptions,
    createDefaultFieldConfig
  );

const configurablePropertiesTest: Array<[string, string | boolean]> = [
  ['title', 'My panel title'],
  ['description', 'My panel description'],
  ['displayMode', 'transparent'],
  ['hoverHeader', true],
];

describe('VizPanelBuilder', () => {
  it.each(configurablePropertiesTest)('should configure %s', (property, expected) => {
    const methodName = `set${property.charAt(0).toUpperCase()}${property.slice(1)}`;
    const builder = getTestBuilder();

    // @ts-ignore
    builder[methodName](expected);

    // @ts-ignore
    expect(builder.build().state[property]).toEqual(expected);
    expect(builder.build().state.pluginId).toEqual('testplugin');
    expect(builder.build().state.pluginVersion).toEqual('1.0.0');
  });

  describe('options and field config', () => {
    it('provides default options', () => {
      const builder = getTestBuilder();
      expect(builder.build().state.options).toEqual(createDefaultOptions());
    });

    it('provides default field config', () => {
      const builder = getTestBuilder();
      expect(builder.build().state.fieldConfig).toEqual({
        defaults: {
          custom: createDefaultFieldConfig(),
        },
        overrides: [],
      });
    });

    describe('standard panels defaults', () => {
      it.each(Object.keys(PanelBuilders))('provides %s defaults', (std) => {
        // @ts-ignore
        const builder = PanelBuilders[std]();

        expect(builder.build().state.options).toMatchSnapshot();
        expect(builder.build().state.fieldConfig).toMatchSnapshot();
      });
    });

    it('allows options configuration', () => {
      const builder = getTestBuilder();
      builder.setOption('complex', { a: 2 }).setOption('numeric', 2);

      expect(builder.build().state.options.complex!.a).toEqual(2);
      expect(builder.build().state.options.complex!.b).toEqual('text');
      expect(builder.build().state.options.numeric).toEqual(2);
      expect(builder.build().state.options.text).toEqual('text');
    });

    it('allows standard field config configuration', () => {
      const builder = getTestBuilder();

      builder
        .setDecimals(2)
        .setUnit('ms')
        .setDisplayName('testDisplayName')
        .setColor({ mode: 'thresholds' })
        .setMin(1)
        .setMax(100)
        .setMappings([
          {
            type: MappingType.ValueToText,
            options: {
              '1': {
                text: 'one',
              },
            },
          },
        ])
        .setThresholds({
          mode: ThresholdsMode.Absolute,
          steps: [
            { color: 'green', value: 0 },
            { color: 'red', value: 80 },
          ],
        })
        .setNoValue('no value')
        .setFilterable(true)
        .setLinks([
          {
            url: 'https://grafana.com',
            title: 'Grafana',
            targetBlank: true,
          },
        ]);

      expect(builder.build().state.fieldConfig!.defaults).toMatchInlineSnapshot(`
        {
          "color": {
            "mode": "thresholds",
          },
          "custom": {
            "complex": {
              "a": 1,
              "b": "text",
            },
            "numeric": 1,
            "text": "text",
          },
          "decimals": 2,
          "displayName": "testDisplayName",
          "filterable": true,
          "links": [
            {
              "targetBlank": true,
              "title": "Grafana",
              "url": "https://grafana.com",
            },
          ],
          "mappings": [
            {
              "options": {
                "1": {
                  "text": "one",
                },
              },
              "type": "value",
            },
          ],
          "max": 100,
          "min": 1,
          "noValue": "no value",
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": 0,
              },
              {
                "color": "red",
                "value": 80,
              },
            ],
          },
          "unit": "ms",
        }
      `);
    });

    it('allows field config configuration', () => {
      const builder = getTestBuilder();

      builder.setCustomFieldConfig('complex', { a: 2 });
      builder.setCustomFieldConfig('numeric', 2);

      expect(builder.build().state.fieldConfig!.defaults.custom).toMatchInlineSnapshot(`
        {
          "complex": {
            "a": 2,
            "b": "text",
          },
          "numeric": 2,
          "text": "text",
        }
      `);
    });

    it('allows multiple panels of the same type configuration', () => {
      const p1 = getTestBuilder().setTitle('p1').build();
      const p2 = getTestBuilder().setTitle('p2').setOption('numeric', 2).build();

      expect(p1.state.options).toMatchInlineSnapshot(`
        {
          "complex": {
            "a": 1,
            "b": "text",
          },
          "numeric": 1,
          "text": "text",
        }
      `);
      expect(p2.state.options).toMatchInlineSnapshot(`
        {
          "complex": {
            "a": 1,
            "b": "text",
          },
          "numeric": 2,
          "text": "text",
        }
      `);
    });
  });

  describe('overrides', () => {
    it('allows standard overrides configuration', () => {
      const builder = getTestBuilder();
      builder.setOverrides((b) =>
        b
          .matchFieldsWithName('fieldName')
          .overrideColor({ mode: 'thresholds' })
          .overrideDecimals(2)
          .overrideDisplayName('testDisplayName')
          .overrideFilterable(true)
          .overrideMax(100)
          .overrideMin(1)
          .overrideNoValue('no value')
          .overrideUnit('ms')
          .overrideLinks([
            {
              url: 'https://grafana.com',
              title: 'Grafana',
              targetBlank: true,
            },
          ])
          .overrideMappings([
            {
              type: MappingType.ValueToText,
              options: {
                '1': {
                  text: 'one',
                },
              },
            },
          ])
          .overrideThresholds({
            mode: ThresholdsMode.Absolute,
            steps: [
              { color: 'green', value: 0 },
              { color: 'red', value: 80 },
            ],
          })
      );

      expect(builder.build().state.fieldConfig.overrides).toHaveLength(1);
      expect(builder.build().state.fieldConfig.overrides).toMatchInlineSnapshot(`
        [
          {
            "matcher": {
              "id": "byName",
              "options": "fieldName",
            },
            "properties": [
              {
                "id": "color",
                "value": {
                  "mode": "thresholds",
                },
              },
              {
                "id": "decimals",
                "value": 2,
              },
              {
                "id": "displayName",
                "value": "testDisplayName",
              },
              {
                "id": "filterable",
                "value": true,
              },
              {
                "id": "max",
                "value": 100,
              },
              {
                "id": "min",
                "value": 1,
              },
              {
                "id": "noValue",
                "value": "no value",
              },
              {
                "id": "unit",
                "value": "ms",
              },
              {
                "id": "links",
                "value": [
                  {
                    "targetBlank": true,
                    "title": "Grafana",
                    "url": "https://grafana.com",
                  },
                ],
              },
              {
                "id": "mappings",
                "value": [
                  {
                    "options": {
                      "1": {
                        "text": "one",
                      },
                    },
                    "type": "value",
                  },
                ],
              },
              {
                "id": "thresholds",
                "value": {
                  "mode": "absolute",
                  "steps": [
                    {
                      "color": "green",
                      "value": 0,
                    },
                    {
                      "color": "red",
                      "value": 80,
                    },
                  ],
                },
              },
            ],
          },
        ]
      `);
    });
    it('allows overrides configuration', () => {
      const builder = getTestBuilder();
      builder.setOverrides((b) =>
        b
          .matchFieldsWithName('fieldName')
          .overrideCustomFieldConfig('complex', { a: 2, b: 'text' })
          .matchFieldsByQuery('A')
          .overrideCustomFieldConfig('numeric', 2)
          .overrideDecimals(2)
      );

      expect(builder.build().state.fieldConfig.overrides).toHaveLength(2);
      expect(builder.build().state.fieldConfig.overrides).toMatchInlineSnapshot(`
        [
          {
            "matcher": {
              "id": "byName",
              "options": "fieldName",
            },
            "properties": [
              {
                "id": "custom.complex",
                "value": {
                  "a": 2,
                  "b": "text",
                },
              },
            ],
          },
          {
            "matcher": {
              "id": "byFrameRefID",
              "options": "A",
            },
            "properties": [
              {
                "id": "custom.numeric",
                "value": 2,
              },
              {
                "id": "decimals",
                "value": 2,
              },
            ],
          },
        ]
      `);
    });
  });
});
