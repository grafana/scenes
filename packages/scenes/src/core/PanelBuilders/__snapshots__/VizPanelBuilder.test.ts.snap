// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VizPanelBuilder options and field config standard panels defaults provides barchart defaults 1`] = `
{
  "barRadius": 0,
  "barWidth": 0.97,
  "fullHighlight": false,
  "groupWidth": 0.7,
  "orientation": "auto",
  "showValue": "auto",
  "stacking": "none",
  "xTickLabelRotation": 0,
  "xTickLabelSpacing": 0,
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides barchart defaults 2`] = `
{
  "defaults": {
    "custom": {
      "fillOpacity": 80,
      "gradientMode": "none",
      "lineWidth": 1,
    },
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides bargauge defaults 1`] = `
{
  "displayMode": "gradient",
  "maxVizHeight": 300,
  "minVizHeight": 16,
  "minVizWidth": 8,
  "namePlacement": "auto",
  "showUnfilled": true,
  "sizing": "auto",
  "valueMode": "color",
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides bargauge defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides datagrid defaults 1`] = `
{
  "selectedSeries": 0,
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides datagrid defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides flamegraph defaults 1`] = `{}`;

exports[`VizPanelBuilder options and field config standard panels defaults provides flamegraph defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides gauge defaults 1`] = `
{
  "minVizHeight": 75,
  "minVizWidth": 75,
  "showThresholdLabels": false,
  "showThresholdMarkers": true,
  "sizing": "auto",
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides gauge defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides geomap defaults 1`] = `
{
  "layers": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides geomap defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides heatmap defaults 1`] = `
{
  "calculate": false,
  "cellGap": 1,
  "cellValues": {},
  "color": {
    "exponent": 0.5,
    "fill": "dark-orange",
    "reverse": false,
    "scheme": "Oranges",
    "steps": 64,
  },
  "exemplars": {
    "color": "rgba(255,0,255,0.7)",
  },
  "filterValues": {
    "le": 1e-9,
  },
  "legend": {
    "show": true,
  },
  "showValue": "auto",
  "tooltip": {
    "mode": "single",
    "showColorScale": false,
    "yHistogram": false,
  },
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides heatmap defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides histogram defaults 1`] = `
{
  "bucketOffset": 0,
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides histogram defaults 2`] = `
{
  "defaults": {
    "custom": {
      "fillOpacity": 80,
      "gradientMode": "none",
      "lineWidth": 1,
    },
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides logs defaults 1`] = `{}`;

exports[`VizPanelBuilder options and field config standard panels defaults provides logs defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides news defaults 1`] = `
{
  "showImage": true,
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides news defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides nodegraph defaults 1`] = `{}`;

exports[`VizPanelBuilder options and field config standard panels defaults provides nodegraph defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides piechart defaults 1`] = `
{
  "displayLabels": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides piechart defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides stat defaults 1`] = `
{
  "colorMode": "value",
  "graphMode": "area",
  "justifyMode": "auto",
  "showPercentChange": false,
  "textMode": "auto",
  "wideLayout": true,
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides stat defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides statetimeline defaults 1`] = `
{
  "alignValue": "left",
  "mergeValues": true,
  "rowHeight": 0.9,
  "showValue": "auto",
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides statetimeline defaults 2`] = `
{
  "defaults": {
    "custom": {
      "fillOpacity": 70,
      "lineWidth": 0,
    },
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides statushistory defaults 1`] = `
{
  "colWidth": 0.9,
  "rowHeight": 0.9,
  "showValue": "auto",
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides statushistory defaults 2`] = `
{
  "defaults": {
    "custom": {
      "fillOpacity": 70,
      "lineWidth": 1,
    },
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides table defaults 1`] = `
{
  "cellHeight": "sm",
  "footer": {
    "countRows": false,
    "reducer": [],
    "show": false,
  },
  "frameIndex": 0,
  "showHeader": true,
  "showTypeIcons": false,
  "sortBy": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides table defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides text defaults 1`] = `
{
  "content": "# Title

For markdown syntax help: [commonmark.org/help](https://commonmark.org/help/)",
  "mode": "markdown",
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides text defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides timeseries defaults 1`] = `{}`;

exports[`VizPanelBuilder options and field config standard panels defaults provides timeseries defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides traces defaults 1`] = `{}`;

exports[`VizPanelBuilder options and field config standard panels defaults provides traces defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides trend defaults 1`] = `{}`;

exports[`VizPanelBuilder options and field config standard panels defaults provides trend defaults 2`] = `
{
  "defaults": {
    "custom": {},
  },
  "overrides": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides xychart defaults 1`] = `
{
  "series": [],
}
`;

exports[`VizPanelBuilder options and field config standard panels defaults provides xychart defaults 2`] = `
{
  "defaults": {
    "custom": {
      "label": "auto",
      "show": "points",
    },
  },
  "overrides": [],
}
`;
